integration/integration.go

```diff
@@ -0,0 +1,184 @@
+	if err != nil {
+		return "", err
+	}
+
+	output := getPodLogs(clientset, p)
+	return output, nil
+}
+
+func getClientSet(configPath string) (*kubernetes.Clientset, error) {
+	config, err := clientcmd.BuildConfigFromFlags("", configPath)
+	if err != nil {
+		return nil, err
+	}
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return clientset, nil
+}
+
+func int32Ptr(i int32) *int32 { return &i }
+
+func findPodForJob(clientset *kubernetes.Clientset, name string, tout, timer time.Duration) (*apiv1.Pod, error) {
+	timeout := time.After(tout)
+	failedPods := make(map[string]struct{})
+	for {
+	podfailed:
+		select {
+		case <-timeout:
+			return nil, fmt.Errorf("podList - time out: no Pod with %s", name)
+		default:
+			pods, err := clientset.CoreV1().Pods(apiv1.NamespaceDefault).List(metav1.ListOptions{})
+			if err != nil {
+				return nil, err
+			}
+			fmt.Printf("Found (%d) pods\n", len(pods.Items))
+			for _, cp := range pods.Items {
+				if _, found := failedPods[cp.Name]; found {
+					continue
+				}
+
+				if strings.HasPrefix(cp.Name, name) {
+					fmt.Printf("pod (%s) - %#v\n", cp.Name, cp.Status.Phase)
+					if cp.Status.Phase == apiv1.PodSucceeded {
+						return &cp, nil
+					}
+
+					if cp.Status.Phase == apiv1.PodFailed {
+						fmt.Printf("pod (%s) - %s - retrying...\n", cp.Name, cp.Status.Phase)
+						failedPods[cp.Name] = struct{}{}
+						break podfailed
+					}
+
+					// Pod still working
+					// Wait and try again...
+					ticker := time.NewTicker(timer)
+					for {
+						fmt.Println("using ticker and an timer...")
+						select {
+						case <-ticker.C:
+							thePod, err := clientset.CoreV1().Pods(apiv1.NamespaceDefault).Get(cp.Name, metav1.GetOptions{})
+							if err != nil {
+								return nil, err
+							}
+							fmt.Printf("thePod (%s) - status:%#v reason:%s message:%s\n", thePod.Name, thePod.Status.Phase, thePod.Status.Reason, thePod.Status.Message)
+							if thePod.Status.Phase == apiv1.PodSucceeded {
+								return thePod, nil
+							}
+
+							if thePod.Status.Phase == apiv1.PodFailed {
+								fmt.Printf("thePod (%s) - %s - retrying...\n", thePod.Name, thePod.Status.Phase)
+								failedPods[thePod.Name] = struct{}{}
+								ticker.Stop()
+								break podfailed
+							}
+
+							if thePod.Status.Phase == apiv1.PodPending && strings.Contains(thePod.Status.Reason, "Failed") {
+								fmt.Printf("thePod (%s) - %s - retrying...\n", thePod.Name, thePod.Status.Reason)
+								failedPods[thePod.Name] = struct{}{}
+								ticker.Stop()
+								break podfailed
+							}
+
+						case <-timeout:
+							ticker.Stop()
+							return nil, fmt.Errorf("getPod time out: no Pod with %s", name)
+						}
+					}
+				}
+			}
+		}
+		time.Sleep(1 * time.Second)
+	}
+
+	return nil, fmt.Errorf("no Pod with %s", name)
+}
+
+func getPodLogs(clientset *kubernetes.Clientset, pod *apiv1.Pod) string {
```

Comment:
What are your thoughts on returning (string, error) in case pod logs were not obtained? The caller of getPodLogs() can then decide appropriately what to do with it.

Reply:
I see your point, but I wanted getPodLogs to be a simple either got logs or string error. Since we're returning that string to the test method. If we decide to use this method for other purposes, we can refactor it to make it better.